src/System.Memory/src/System/Buffers/ReadOnlySequence_helpers.cs

```diff
@@ -467,5 +467,81 @@ private static bool InRange(ulong value, ulong start, ulong end)
             // Equivalent to: return (start <= value && value <= start)
             return (value - start) <= (end - start);
         }
+
+        /// <summary>
+        /// Helper to efficiently prepare the <see cref="SequenceReader{T}"/>
+        /// </summary>
+        /// <param name="first">The first span in the sequence.</param>
+        /// <param name="next">The next position.</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal void GetFirstSpan(out ReadOnlySpan<T> first, out SequencePosition next)
+        {
+            first = default;
+            next = default;
+            SequencePosition start = Start;
+            int startIndex = start.GetInteger();
+            object startObject = start.GetObject();
+
+            if (startObject != null)
+            {
+                SequencePosition end = End;
+                int endIndex = end.GetInteger();
+                bool isMultiSegment = startObject != end.GetObject();
+
+                // A == 0 && B == 0 means SequenceType.MultiSegment
+                if (startIndex >= 0)
+                {
+                    if (endIndex >= 0)  // SequenceType.MultiSegment
+                    {
+                        ReadOnlySequenceSegment<T> segment = (ReadOnlySequenceSegment<T>)startObject;
+                        next = new SequencePosition(segment.Next, 0);
+                        first = segment.Memory.Span;
+                        if (isMultiSegment)
+                        {
+                            first = first.Slice(startIndex);
+                        }
+                        else
+                        {
+                            first = first.Slice(startIndex, endIndex - startIndex);
+                        }
+                    }
+                    else
+                    {
+                        if (isMultiSegment)
+                            ThrowHelper.ThrowInvalidOperationException_EndPositionNotReached();
+
+                        first = new ReadOnlySpan<T>((T[])startObject, startIndex, (endIndex & ReadOnlySequence.IndexBitMask) - startIndex);
+                    }
+                }
+                else
+                {
+                    first = GetFirstSpanSlow(startObject, startIndex, endIndex, isMultiSegment);
+                }
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private static ReadOnlySpan<T> GetFirstSpanSlow(object startObject, int startIndex, int endIndex, bool isMultiSegment)
+        {
+            Debug.Assert(startIndex < 0 || endIndex < 0);
+            if (isMultiSegment)
+                ThrowHelper.ThrowInvalidOperationException_EndPositionNotReached();
+
+            // The type == char check here is redundant. However, we still have it to allow
+            // the JIT to see when that the code is unreachable and eliminate it.
+            // A == 1 && B == 1 means SequenceType.String
```

Comment:
What are `A` and `B` in these comments?

Reply:
Er, copy/paste fail. I'll update the comment.